crates/matrix-sdk/src/room/timeline/inner.rs

```diff
@@ -163,39 +163,42 @@ impl<P: ProfileProvider> TimelineInner<P> {
     /// server, so the sending has failed. If the event ID is `Some(_)`, it
     /// means the sending has been successful.
     ///
-    /// If no local event is found, a warning is raised.   
+    /// If no local event is found, a warning is raised.
     pub(super) fn update_event_id_of_local_event(
         &self,
         txn_id: &TransactionId,
         event_id: Option<OwnedEventId>,
     ) {
         let mut lock = self.items.lock_mut();
 
-        // Look for the local event by the transaction ID.
-        if let Some((idx, local_event_item)) = find_event_by_txn_id(&lock, txn_id) {
-            // An event ID already exists, that's a broken state, let's emit an error but
-            // also override to the given event ID.
-            if let Some(existing_event_id) = &local_event_item.event_id {
-                error!(
-                    ?existing_event_id, new_event_id = ?event_id, ?txn_id,
-                    "Local echo already has an event ID"
-                );
-            }
+        // Look for the local event by the transaction ID or event ID.
+        let result = rfind_event_item(&lock, |it| {
+            it.transaction_id() == Some(txn_id)
+                || event_id.is_some() && it.event_id() == event_id.as_deref()
+        });
```

Comment:
I've concerned this code might not be optimal. It will run in a loop. `it.transaction_id()` and `it.event_id()` both will do a `match` to detect whether it's a local or a remote event, same for `event_id.as_deref()` which will do a `match` on `Some` or `None`.
Well, on the other hand, it's more likely that we are looking for a local-echo, so… the first part of the condition will break/return early. Just thinking out loud.

Reply:
Maybe (not tested, written directly in Github):

```suggestion
        let result = rfind_event_item(&lock, |it| {
            match it {
                EventTimelineItem::Local(LocalEventTimelineItem { transaction_id, event_id: ev_id, .. }) => transaction_id == txn_id || ev_id == event_id,
                EventTimelineItem::Remote(_) => false,
            }
        });
```

It can even be written as such:

```suggestion
    let result = rfind_event_item(&lock, |it| {
        matches!(it, EventTimelineItem::Local(LocalEventTimelineItem { transaction_id, event_id: ev_id, .. }) if transaction_id == txn_id || ev_id == event_id);
    });
```

Thoughts?

Reply:
Regarding runtime performance, the optimal thing would probably be to have two branches that use a different closure for searching. But I doubt the matching logic inside the search is going to kill performance. Or if it does, we should introduce txn-id to index and event-id to index maps.